docs/blog/2021-12-01-csp.md

```diff
@@ -0,0 +1,208 @@
+    }
+);
+```
+
+Go ahead and
+try [the url from that script](<https://accounts.google.com/o/oauth2/revoke?callback=alert('This is a malicious alert')>)
+in your browser. You'll quickly see the problem. Now an attacker that can embed HTML in your page has the ability to
+execute whatever JavaScript by having the server reflect it back to your page. Any un-sanitized field that leaks HTML
+into your page is now an immediate vulnerability.
+
+There a great number of these vulnerable legacy endpoints, not just on Google's domains but across the web.  
+Here is a list of [just some of them](https://github.com/zigoo0/JSONBee/blob/master/jsonp.txt).
+
+Because this isn't a specific protocol but just a serverside (anti)pattern, there's no way for the browser to detect
+this behavior and disable it. The only solution is to not allow scripts to load from any vulnerable domain, and to host
+them yourself instead.
+
+**The bigger your website, the harder this is to deal with.**
+
+If you have a small site and immediate control over the codebase, this could be an easy problem to patch.
+
+For large sites at big companies that load analytics payloads, third party auth, and other scripts from these vulnerable
+domains, it can be difficult to mitigate. Heck, a lot of big companies have legacy JSONP endpoints on the _same domain
+as the site._  There's not much you can do about that.
+
+### File Upload Bypass
+
+Let's say your CSP is locked to only allow scripts from your domain.
+
+If your site allows file uploads and downloads and doesn't discriminate against files that end in `.js` with content
+type `application/javascript`. A user can simply upload a file `malicious-code.js`, and load it as a script tag,
+directly from your domain.  
+
+```html
+
+<script src=www.yourdomain.com/myfiles/malicious-code.js></script>
+```
+
+You have to be very careful about what your users can upload.  In some cases the browser will try to automatically detect
+and fix MIME types in a process called sniffing, which could convert something you didn't block back into `application/javascript` during download.
+File downloads need to [include the header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options) `X-Content-Type-Options: nosniff` to prevent that.  
+
+As you can see, these are nuanced details that are easy to miss.
+
+### Other vulnerabilities
+
+These are just a few of the potential ways to bypass the CSP's protections. The
+list [goes on](https://book.hacktricks.xyz/pentesting-web/content-security-policy-csp-bypass).
+
+### Mitigations
+
+For large sites, there may be just too many gaps in the armor to treat the CSP as a functional line of defense, and
+many big sites have disabled it entirely.
+[This paper](https://research.google/pubs/pub45542/) found that of those that use it **over 94% of sites have
+ineffective CSPs!**
+
+Long story short, there's a good chance your CSP isn't doing much to protect your site from XSS.
```

Comment:
"The only effective way to enforce CSP is by abiding by what is known as 'strict csp'" https://csp.withgoogle.com/docs/strict-csp.html